File: src/main/java/dev/nytweetdeck/xapi/graphql/AuthenticatedGraphQlClient.java
package dev.nytweetdeck.xapi.graphql;

import dev.nytweetdeck.account.AccountStore;
import dev.nytweetdeck.xapi.http.XApiHttpException;
import dev.nytweetdeck.xapi.http.WebSessionRequestHeaders;
import dev.nytweetdeck.xapi.http.XClientTransactionIdService;
import dev.nytweetdeck.xapi.profile.XApiProfile;
import dev.nytweetdeck.xapi.profile.XApiProfile.OperationType;
import dev.nytweetdeck.xapi.profile.XApiProfileService;
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.ObjectMapper;

@Service
public class AuthenticatedGraphQlClient {

    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(30);
    private static final Pattern LANGUAGE_PATTERN =
            Pattern.compile("[A-Za-z]{2,3}(?:-[A-Za-z0-9]{2,8})*");
    private static final Set<String> POST_QUERY_PURPOSES = Set.of("search");
    private static final Map<String, Boolean> DEFAULT_FIELD_TOGGLES = Map.of(
            "withPayments", false,
            "withAuxiliaryUserLabels", false,
            "withArticleRichContentState", false,
            "withArticlePlainText", false,
            "withArticleSummaryText", false,
            "withArticleVoiceOver", false,
            "withGrokAnalyze", false,
            "withDisallowedReplyControls", false);

    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final XApiProfileService profileService;
    private final AccountStore accountStore;
    private final XClientTransactionIdService transactionIdService;

    @Autowired
    public AuthenticatedGraphQlClient(
            HttpClient httpClient,
            ObjectMapper objectMapper,
            XApiProfileService profileService,
            AccountStore accountStore,
            XClientTransactionIdService transactionIdService) {
        this.httpClient = httpClient;
        this.objectMapper = objectMapper;
        this.profileService = profileService;
        this.accountStore = accountStore;
        this.transactionIdService = transactionIdService;
    }

    public AuthenticatedGraphQlClient(
            HttpClient httpClient,
            ObjectMapper objectMapper,
            XApiProfileService profileService,
            AccountStore accountStore) {
        this(httpClient, objectMapper, profileService, accountStore, null);
    }

    public GraphQlResult execute(
            String accountId, String purpose, Map<String, Object> variables) {
        return execute(accountId, purpose, variables, "ja");
    }

    public GraphQlResult execute(
            String accountId,
            String purpose,
            Map<String, Object> variables,
            String language) {
        for (int attempt = 0; attempt < 2; attempt++) {
            var prepared = prepareRequest(accountId, purpose, variables, language);
            try {
                var response = httpClient.send(
                        prepared.request(), HttpResponse.BodyHandlers.ofString());
                if (shouldRefreshTransaction(prepared, response, attempt)) {
                    transactionIdService.invalidate();
                    continue;
                }
                if (response.statusCode() < 200 || response.statusCode() >= 300) {
                    throw new XApiHttpException(
                            "GraphQL " + purpose + "に失敗しました。HTTP " + response.statusCode(),
                            response.statusCode());
                }
                rejectGraphQlErrors(response.body(), purpose);
                return new GraphQlResult(
                        purpose, prepared.operation().operationName(), response.body());
            } catch (IOException exception) {
                throw new XApiHttpException("GraphQL " + purpose + "の通信に失敗しました。", exception);
            } catch (InterruptedException exception) {
                Thread.currentThread().interrupt();
                throw new XApiHttpException("GraphQL " + purpose + "が中断されました。", exception);
            }
        }
        throw new XApiHttpException("GraphQL " + purpose + "のWeb署名を更新できませんでした。", 502);
    }

    PreparedRequest prepareRequest(
            String accountId, String purpose, Map<String, Object> variables) {
        return prepareRequest(accountId, purpose, variables, "ja");
    }

    PreparedRequest prepareRequest(
            String accountId,
            String purpose,
            Map<String, Object> variables,
            String language) {
        var profile = profileService.profile();
        var operation = profileService.requireOperation(purpose);
        var features = profileService.selectFeatures(operation);
        var account = accountStore.requireAccount(accountId);
        var operationUri = operation.resolveAgainst(profile.graphqlBaseUri());

        var fieldToggles = operation.fieldToggles().isEmpty()
                ? DEFAULT_FIELD_TOGGLES
                : operation.fieldToggles().stream()
                        .collect(java.util.stream.Collectors.toUnmodifiableMap(
                                key -> key, key -> false));
        String body = null;
        URI requestUri;
        if (operation.type() == OperationType.MUTATION) {
            var payload = new LinkedHashMap<String, Object>();
            payload.put("variables", variables);
            payload.put("queryId", operation.operationId());
            if (!operation.featureKeys().isEmpty()) {
                payload.put("features", features);
            }
            body = writeJson(payload);
            requestUri = operationUri;
        } else if (POST_QUERY_PURPOSES.contains(purpose)) {
            body = postQueryBody(variables, features, fieldToggles);
            requestUri = operationUri;
        } else {
            requestUri = withQuery(
                    operationUri,
                    Map.of(
                            "variables", writeJson(variables),
                            "features", writeJson(features),
                            "fieldToggles", writeJson(fieldToggles)));
        }

        var requestBuilder = HttpRequest.newBuilder(requestUri).timeout(REQUEST_TIMEOUT);
        WebSessionRequestHeaders.apply(requestBuilder, account, normalizeLanguage(language));
        if (body == null) {
            requestBuilder.GET();
        } else {
            requestBuilder
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body));
        }
        if (operation.type() == OperationType.MUTATION) {
            if (transactionIdService == null) {
                throw new IllegalStateException("X Web署名サービスが接続されていません。");
            }
            requestBuilder.header(
                    "X-Client-Transaction-Id",
                    transactionIdService.generate("POST", requestUri));
        }
        return new PreparedRequest(requestBuilder.build(), operation);
    }

    private boolean shouldRefreshTransaction(
            PreparedRequest prepared,
            HttpResponse<String> response,
            int attempt) {
        return attempt == 0
                && transactionIdService != null
                && prepared.operation().type() == OperationType.MUTATION
                && (response.statusCode() == 404 || hasGraphQlErrorCode(response.body(), 344));
    }

    private boolean hasGraphQlErrorCode(String body, int expectedCode) {
        try {
            var errors = objectMapper.readTree(body).get("errors");
            if (errors == null || !errors.isArray()) {
                return false;
            }
            for (var error : errors) {
                var code = error.get("code");
                if (code != null && code.canConvertToInt() && code.asInt() == expectedCode) {
                    return true;
                }
            }
            return false;
        } catch (JacksonException exception) {
            return false;
        }
    }

    private void rejectGraphQlErrors(String body, String purpose) {
        try {
            var root = objectMapper.readTree(body);
            var errors = root.get("errors");
            var data = root.get("data");
            if (errors != null
                    && errors.isArray()
                    && !errors.isEmpty()
                    && (data == null || data.isNull() || data.isEmpty())) {
                throw new XApiHttpException(
                        "GraphQL " + purpose + "がエラーを返しました。", 502);
            }
        } catch (JacksonException exception) {
            throw new XApiHttpException("GraphQL応答を解析できません。", exception);
        }
    }

    private String writeJson(Object value) {
        try {
            return objectMapper.writeValueAsString(value);
        } catch (JacksonException exception) {
            throw new IllegalArgumentException("GraphQL入力をJSONへ変換できません。", exception);
        }
    }

    private String postQueryBody(
            Map<String, Object> variables,
            Map<String, Boolean> features,
            Map<String, Boolean> fieldToggles) {
        return writeJson(Map.of(
                "variables", variables,
                "features", features,
                "fieldToggles", fieldToggles));
    }

    private static URI withQuery(URI uri, Map<String, String> values) {
        var sorted = new LinkedHashMap<String, String>();
        values.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .forEach(entry -> sorted.put(entry.getKey(), entry.getValue()));
        var query = String.join("&", sorted.entrySet().stream()
                .map(entry -> encode(entry.getKey()) + "=" + encode(entry.getValue()))
                .toList());
        return URI.create(uri + "?" + query);
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private static String normalizeLanguage(String language) {
        var normalized = language == null ? "" : language.strip().replace('_', '-');
        if (!LANGUAGE_PATTERN.matcher(normalized).matches()) {
            throw new IllegalArgumentException("表示言語の形式が不正です。");
        }
        return normalized.toLowerCase(Locale.ROOT);
    }

    public record GraphQlResult(String purpose, String operationName, String rawJson) {

        @Override
        public String toString() {
            return "GraphQlResult[purpose="
                    + purpose
                    + ", operationName="
                    + operationName
                    + ", rawJson=<redacted>]";
        }
    }

    record PreparedRequest(HttpRequest request, XApiProfile.GraphQlOperation operation) {}
}

File: android/app/src/main/java/dev/nytweetdeck/android/xapi/AuthenticatedGraphQlClient.kt
package dev.nytweetdeck.android.xapi

import java.util.Locale
import java.util.concurrent.TimeUnit
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

data class XSessionCredentials(
    val bearerToken: String,
    val authToken: String,
    val csrfToken: String,
)

fun interface GraphQlExecutor {
    fun execute(
        credentials: XSessionCredentials,
        purpose: String,
        variables: Map<String, Any?>,
        language: String,
    ): String
}

class AuthenticatedGraphQlClient(
    client: OkHttpClient,
    private val profileProvider: () -> XApiProfile,
    private val userAgent: String,
    private val transactionIdService: XClientTransactionIdService =
        XClientTransactionIdService(client, userAgent),
) : GraphQlExecutor {
    constructor(
        client: OkHttpClient,
        profile: XApiProfile,
        userAgent: String,
    ) : this(client, { profile }, userAgent)

    constructor(
        client: OkHttpClient,
        profile: XApiProfile,
        userAgent: String,
        transactionIdService: XClientTransactionIdService,
    ) : this(client, { profile }, userAgent, transactionIdService)

    private val client = client.newBuilder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .callTimeout(35, TimeUnit.SECONDS)
        .build()

    override fun execute(
        credentials: XSessionCredentials,
        purpose: String,
        variables: Map<String, Any?>,
        language: String,
    ): String {
        validateCredentials(credentials)
        val normalizedLanguage = normalizeLanguage(language)
        val profile = profileProvider()
        val operation = profile.requireOperation(purpose)
        val features = profile.featuresFor(operation)
        val toggles = selectFieldToggles(purpose, operation)
        val base = profile.graphqlBaseUrl.toHttpUrl().newBuilder()
            .addPathSegment(operation.operationId)
            .addPathSegment(operation.operationName)
            .build()
        val variablesJson = mapToJson(variables)
        val featuresJson = mapToJson(features)
        val togglesJson = mapToJson(toggles)

        val requestBuilder = Request.Builder()
            .header("Authorization", "Bearer ${credentials.bearerToken}")
            .header("Cookie", "auth_token=${credentials.authToken}; ct0=${credentials.csrfToken}")
            .header("X-CSRF-Token", credentials.csrfToken)
            .header("X-Twitter-Auth-Type", "OAuth2Session")
            .header("X-Twitter-Active-User", "yes")
            .header("X-Twitter-Client-Language", normalizedLanguage)
            .header("Accept-Language", normalizedLanguage)
            .header("Origin", "https://x.com")
            .header("Referer", "https://x.com/")
            .header("User-Agent", userAgent)
            .header("Accept", "application/json")

        when {
            operation.type == XApiProfile.OperationType.MUTATION -> {
                val body = buildJsonObject {
                    put("variables", variablesJson)
                    put("queryId", operation.operationId)
                    if (operation.featureKeys.isNotEmpty()) {
                        put("features", featuresJson)
                    }
                }.toString().toRequestBody(JSON_MEDIA_TYPE)
                requestBuilder.url(base).post(body)
            }
            purpose == "search" -> {
                val body = buildJsonObject {
                    put("variables", variablesJson)
                    put("features", featuresJson)
                    put("fieldToggles", togglesJson)
                }.toString().toRequestBody(JSON_MEDIA_TYPE)
                requestBuilder.url(base).post(body)
            }
            else -> {
                val url = base.newBuilder()
                    .addQueryParameter("features", featuresJson.toString())
                    .addQueryParameter("fieldToggles", togglesJson.toString())
                    .addQueryParameter("variables", variablesJson.toString())
                    .build()
                requestBuilder.url(url).get()
            }
        }

        val unsignedRequest = requestBuilder.build()
        for (attempt in 0 until MAX_TRANSACTION_ATTEMPTS) {
            val request = withTransactionHeader(unsignedRequest, "GraphQL ${purpose}")
            val response = executeRequest(request, purpose)
            if (isSignatureRejected(request, response)) {
                if (attempt == 0) {
                    transactionIdService.invalidate()
                    continue
                }
                throw XApiException("GraphQL ${purpose}のWeb署名を更新できませんでした。", 502)
            }
            if (response.statusCode !in 200..299) {
                throw XApiException("GraphQL ${purpose}に失敗しました。HTTP ${response.statusCode}", response.statusCode)
            }
            rejectTerminalGraphQlErrors(response.body, purpose)
            return response.body
        }
        throw XApiException("GraphQL ${purpose}のWeb署名を更新できませんでした。", 502)
    }

    private fun withTransactionHeader(request: Request, scope: String): Request {
        if (!XClientTransactionIdService.supportsOfficialApiRequest(request.url)) {
            return request
        }
        val transactionId = generateTransactionId(request, scope)
        return request.newBuilder().header(TRANSACTION_HEADER, transactionId).build()
    }

    private fun generateTransactionId(request: Request, scope: String): String {
        var lastFailure: Exception? = null
        for (attempt in 0 until MAX_TRANSACTION_ATTEMPTS) {
            try {
                return transactionIdService.generate(request.method, request.url)
            } catch (exception: Exception) {
                lastFailure = exception
                if (attempt == 0) {
                    transactionIdService.invalidate()
                }
            }
        }
        throw XApiException("${scope}のWeb署名を生成できません。", 502, lastFailure)
    }

    private fun executeRequest(request: Request, purpose: String): ApiResponse = try {
        client.newCall(request).execute().use { response ->
            ApiResponse(response.code, response.body.string())
        }
    } catch (exception: XApiException) {
        throw exception
    } catch (exception: Exception) {
        throw XApiException("GraphQL ${purpose}の通信に失敗しました。", cause = exception)
    }

    private fun isSignatureRejected(
        request: Request,
        response: ApiResponse,
    ): Boolean =
        request.header(TRANSACTION_HEADER) != null &&
        (response.statusCode == 404 || hasGraphQlErrorCode(response.body, SIGNATURE_ERROR_CODE))

    private fun hasGraphQlErrorCode(body: String, expectedCode: Int): Boolean = try {
        val errors = Json.parseToJsonElement(body).jsonObject["errors"] as? JsonArray ?: return false
        errors.any { error ->
            ((error as? JsonObject)?.get("code") as? JsonPrimitive)?.intOrNull == expectedCode
        }
    } catch (_: Exception) {
        false
    }

    private fun rejectTerminalGraphQlErrors(body: String, purpose: String) {
        val root = try {
            Json.parseToJsonElement(body).jsonObject
        } catch (exception: Exception) {
            throw XApiException("GraphQL応答を解析できません。", cause = exception)
        }
        val errors = root["errors"] as? JsonArray
        val data = root["data"]
        if (!errors.isNullOrEmpty() && (data == null || data is JsonNull || data == JsonObject(emptyMap()))) {
            throw XApiException("GraphQL ${purpose}がエラーを返しました。", 502)
        }
    }

    private fun selectFieldToggles(
        purpose: String,
        operation: XApiProfile.GraphQlOperation,
    ): Map<String, Boolean> {
        val keys = operation.fieldToggles.ifEmpty { DEFAULT_FIELD_TOGGLES.keys.toList() }
        val includeArticle = purpose == "postDetail" || purpose == "conversation"
        return keys.associateWith { key ->
            when (key) {
                "withArticleRichContentState", "withArticlePlainText" -> includeArticle
                "withArticleSummaryText" -> true
                else -> DEFAULT_FIELD_TOGGLES[key] ?: false
            }
        }
    }

    private fun normalizeLanguage(value: String): String {
        val normalized = value.trim().replace('_', '-')
        require(LANGUAGE_PATTERN.matches(normalized)) { "表示言語の形式が不正です。" }
        return normalized.lowercase(Locale.ROOT)
    }

    private fun validateCredentials(credentials: XSessionCredentials) {
        require(credentials.bearerToken.isNotBlank()) { "Web Bearerが空です。" }
        require(credentials.authToken.isNotBlank()) { "auth_tokenが空です。" }
        require(credentials.csrfToken.isNotBlank()) { "ct0が空です。" }
    }

    private fun mapToJson(values: Map<String, Any?>): JsonObject = JsonObject(
        values.mapValues { (_, value) -> value.toJsonElement() },
    )

    private fun Any?.toJsonElement(): JsonElement = when (this) {
        null -> JsonNull
        is Boolean -> JsonPrimitive(this)
        is Number -> JsonPrimitive(this)
        is String -> JsonPrimitive(this)
        is Map<*, *> -> JsonObject(entries.associate { (key, value) ->
            require(key is String) { "GraphQL map keyは文字列である必要があります。" }
            key to value.toJsonElement()
        })
        is Iterable<*> -> JsonArray(map { it.toJsonElement() })
        else -> throw IllegalArgumentException("GraphQL変数の型が不正です。")
    }

    private companion object {
        const val MAX_TRANSACTION_ATTEMPTS = 2
        const val SIGNATURE_ERROR_CODE = 344
        const val TRANSACTION_HEADER = "X-Client-Transaction-Id"
        val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
        val LANGUAGE_PATTERN = Regex("[A-Za-z]{2,3}(?:-[A-Za-z0-9]{2,8})*")
        val DEFAULT_FIELD_TOGGLES = mapOf(
            "withPayments" to false,
            "withAuxiliaryUserLabels" to false,
            "withArticleRichContentState" to false,
            "withArticlePlainText" to false,
            "withArticleSummaryText" to true,
            "withArticleVoiceOver" to false,
            "withGrokAnalyze" to false,
            "withDisallowedReplyControls" to false,
        )
    }

    private data class ApiResponse(
        val statusCode: Int,
        val body: String,
    )
}
